Migrate CRI tests CI to bare metal

CuriousGeorgiy opened the pull request:
## Summary

Closes #705

## Implementation Notes :hammer_and_pick:

Currently, we use kind and Docker to setup an environment for the CI
workflow for CRI tests: this approach is hard to maintain because any
dependency update requires setting up a new Docker image and resolving
dependency conflicts and compatibility — let's use a bare metal setup
instead.

Refactor CRI tests workflow:
* create a callable workflow parametrized by sandbox;
* create separate workflows for each sandbox that reuse the callable
  workflow.

Drop redundant environment variables from CRI tests workflow:
* GOOS default is suitable;
* TMPDIR default is suitable;
* GOCACHE default is suitable;
* GOPATH default is suitable;

Drop redundant steps from CRI tests workflow:
* `/tmp` directory always exists;
* go is already installed during node setup;
* `/etc/vhive-cri` is already created during node setup;
* sandbox is already setup during node setup;
* `go build` is run during CRI test environment setup.

Merge Makefile in `./cri` and
`scripts_github_runner/setup_cri_dev_env.sh` to
`scripts/github_runner/setup_cri_test_env.sh`, dropping the following
from the latter:
* protobuf setup — already done during node setup;
* system setup — already done during node setup;
* exec — this script was originally run as a docker container entrypoint.

Drop docker files for setting up CRI test environment and GitHub runner
using docker.

Drop the following redundant scripts:
* `scripts/github_runner/setup_runner.sh`
* `scripts/github_runner/setup_runner_cri.sh`
— they basically did the same as
`scripts/github_runner/setup_bare_metal_runner.sh` but were adopted to
run inside docker.

Drop 'cri' and 'gvisor-cri' labels from `start_runner.sh` since this script
only works with a kind-docker setup (i.e., it spawns a specified number of
runners in docker containers and, if specified, restarts them).

Refactor `scripts/github_runner/start_bare-metal_runner.sh` script:
* rename it properly;
* adjust it to being run from an already cloned vhive repository.

Add a section about running CRI tests locally to developers guide.

## External Dependencies :four_leaf_clover:

N/A.

## Breaking API Changes :warning:

N/A.

anshalshukla commented on .github/workflows/cri_tests.yml:
```diff
@@ -0,0 +1,47 @@
+        with:
+          name: ctrd-logs
+          path: |
+            /tmp/ctrd-logs/${{ github.run_id }}
```
tmp directory is not setup, you should add back that step

CuriousGeorgiy replied:
AFAIC, `/tmp` directory always exists on unix systems.

anshalshukla commented on docs/quickstart_guide.md:
````diff
@@ -152,6 +152,7 @@ SSD-equipped nodes are highly recommended. Full list of CloudLab nodes can be fo
 3. Start `firecracker-containerd` in a background terminal named `firecracker`:
     ```bash
     sudo PATH=$PATH screen -dmS firecracker bash -c "/usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml > >(tee -a /tmp/vhive-logs/firecracker.stdout) 2> >(tee -a /tmp/vhive-logs/firecracker.stderr >&2)"
+    sudo PATH=$PATH screen -dmS gvisor bash -c "/usr/local/bin/gvisor-containerd --address /run/gvisor-containerd/gvisor-containerd.sock --config /etc/gvisor-containerd/config.toml > >(tee -a /tmp/vhive-logs/gvisor.stdout) 2> >(tee -a /tmp/vhive-logs/gvisor.stderr >&2)"
````
I think this will lead to confusion, in my opinion a separate section/ sub-section should be created for gvisor like there is for stock-only. This will be misleading for newcomers as the point above mentions that the script can be used to start firecracker-containerd

ustiugov replied:
I agree with Anshal, need a separate note.

anshalshukla commented on docs/quickstart_guide.md:
````diff
@@ -161,7 +162,8 @@ SSD-equipped nodes are highly recommended. Full list of CloudLab nodes can be fo
 5. Start `vHive` in a background terminal named `vhive`:
     ```bash
     # EITHER
-    sudo screen -dmS vhive bash -c "./vhive > >(tee -a /tmp/vhive-logs/vhive.stdout) 2> >(tee -a /tmp/vhive-logs/vhive.stderr >&2)"
````
Again this will be confusing, also I'm not sure if this command should be deleted. I think it is there to run vhive in various modes with the firecracker

CuriousGeorgiy replied:
These are my tweaks that got into the diff unintentionally, dropped them.

anshalshukla commented on scripts/github_runner/setup_cri_test_env.sh:
```diff
@@ -22,31 +22,28 @@
 
-sudo sysctl --system
-sudo swapoff -a
+source etc/profile && go run $ROOT/examples/registry/populate_registry.go -imageFile $ROOT/examples/registry/images.txt
```
we should not have these untill we update the ubuntu version in vhive profile of cloudlab, local-registry tests can be skipped for now. I've create a separate issue for that you can comment this line and reference it here

ustiugov replied:
@anshalshukla let's merge the PR merging the local-registry removal before this PR and rebase.

CuriousGeorgiy replied:
Opened #710 to close #703.

anshalshukla commented on scripts/github_runner/setup_cri_test_env.sh:
```diff
@@ -22,31 +22,28 @@
-# we want the command (expected to be systemd) to be PID1, so exec to it
-exec "$@"
+KUBECONFIG=/etc/kubernetes/admin.conf sudo kn service apply helloworld -f $ROOT/configs/knative_workloads/helloworld.yaml
+KUBECONFIG=/etc/kubernetes/admin.conf sudo kn service apply helloworldlocal -f $ROOT/configs/knative_workloads/helloworld_local.yaml
```
Same, local registry test

CuriousGeorgiy replied:
Opened #710 to close #703.

anshalshukla commented on scripts/github_runner/setup_bare_metal_runner.sh:
```diff
@@ -22,19 +22,25 @@
+SANDBOX=$5
 
-sudo apt-get update
+ROOT="./../.."
```
I feel relative paths can be confusing, can we have a absolute path here?

ustiugov replied:
IMO, it's better as is.

anshalshukla commented on scripts/github_runner/setup_cri_test_env.sh:
```diff
@@ -22,31 +22,28 @@
-$PWD/create_devmapper.sh
-
-# install Protocol Buffer Compiler
-PROTO_VERSION=3.11.4
```
I guess these are used by CRI not sure though, are you sure we don't need them?

ustiugov reviewed (changes requested):
Thank you! A lot of good work here. 

- [ ] Please add a detailed description of changes to this PR description. Do not forget to to carefully document all the legacy code removals
- [ ] How someone can test CRI locally is unclear. Please document, e.g., in the Dev guide doc. 
- [ ] Some setup script changes are unclear but probably would be more intuitive after the above points are done.

CuriousGeorgiy commented:
> Some setup script changes are unclear but probably would be more intuitive after the above points are done.

@ustiugov if there are any unclear changes or script changes left, please point them out.

ustiugov reviewed (commented):
LGTM

have you tested the "integ" runner option after you have removed the docker files. can we still build them? can you please check?

@shyamjesal please review as well.

CuriousGeorgiy commented:
> have you tested the "integ" runner option after you have removed the docker files. can we still build them? can you please check?

Apparently, the integration tests workflow runner does not depend on the docker files in the repository: https://github.com/vhive-serverless/vHive/blob/c679951297e3456f54f85e86dc59fda63b66e534/scripts/github_runner/start_runners.sh#L56

I followed the steps from the continuous integration guide[^1]:
<details> <summary> ./scripts/github_runner/setup_runner_host.sh </summary>

```console
Get:1 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Get:2 http://us.archive.ubuntu.com/ubuntu focal InRelease [265 kB]
Get:3 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages [2,080 kB]
Get:4 http://us.archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Get:5 http://us.archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]
Get:6 http://us.archive.ubuntu.com/ubuntu focal/main i386 Packages [718 kB]
Get:7 http://security.ubuntu.com/ubuntu focal-security/main i386 Packages [574 kB]
Get:8 http://security.ubuntu.com/ubuntu focal-security/main Translation-en [338 kB]
Get:9 http://security.ubuntu.com/ubuntu focal-security/main amd64 c-n-f Metadata [12.5 kB]
Get:10 http://security.ubuntu.com/ubuntu focal-security/restricted amd64 Packages [1,605 kB]
Get:11 http://security.ubuntu.com/ubuntu focal-security/restricted i386 Packages [29.2 kB]
Get:12 http://security.ubuntu.com/ubuntu focal-security/restricted Translation-en [227 kB]
Get:13 http://security.ubuntu.com/ubuntu focal-security/restricted amd64 c-n-f Metadata [620 B]
Get:14 http://security.ubuntu.com/ubuntu focal-security/universe i386 Packages [590 kB]
Get:15 http://us.archive.ubuntu.com/ubuntu focal/main amd64 Packages [970 kB]
Get:16 http://security.ubuntu.com/ubuntu focal-security/universe amd64 Packages [818 kB]
Get:17 http://us.archive.ubuntu.com/ubuntu focal/main Translation-en [506 kB]
Hit:18 http://us.archive.ubuntu.com/ubuntu focal/main amd64 c-n-f Metadata
Get:19 http://us.archive.ubuntu.com/ubuntu focal/restricted i386 Packages [8,112 B]
Get:20 http://us.archive.ubuntu.com/ubuntu focal/restricted amd64 Packages [22.0 kB]
Get:21 http://us.archive.ubuntu.com/ubuntu focal/restricted Translation-en [6,212 B]
Hit:22 http://us.archive.ubuntu.com/ubuntu focal/restricted amd64 c-n-f Metadata
Get:23 http://security.ubuntu.com/ubuntu focal-security/universe Translation-en [164 kB]
Get:24 http://us.archive.ubuntu.com/ubuntu focal/universe amd64 Packages [8,628 kB]
Get:25 http://security.ubuntu.com/ubuntu focal-security/universe amd64 c-n-f Metadata [17.5 kB]
Get:26 http://us.archive.ubuntu.com/ubuntu focal/universe i386 Packages [4,642 kB]
Get:27 http://us.archive.ubuntu.com/ubuntu focal/universe Translation-en [5,124 kB]
Hit:28 http://us.archive.ubuntu.com/ubuntu focal/universe amd64 c-n-f Metadata
Get:29 http://us.archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages [2,461 kB]
Get:30 http://us.archive.ubuntu.com/ubuntu focal-updates/main i386 Packages [803 kB]
Get:31 http://us.archive.ubuntu.com/ubuntu focal-updates/main Translation-en [420 kB]
Get:32 http://us.archive.ubuntu.com/ubuntu focal-updates/main amd64 c-n-f Metadata [16.4 kB]
Get:33 http://us.archive.ubuntu.com/ubuntu focal-updates/restricted i386 Packages [30.5 kB]
Get:34 http://us.archive.ubuntu.com/ubuntu focal-updates/restricted amd64 Packages [1,713 kB]
Get:35 http://us.archive.ubuntu.com/ubuntu focal-updates/restricted Translation-en [242 kB]
Get:36 http://us.archive.ubuntu.com/ubuntu focal-updates/restricted amd64 c-n-f Metadata [620 B]
Get:37 http://us.archive.ubuntu.com/ubuntu focal-updates/universe i386 Packages [720 kB]
Get:38 http://us.archive.ubuntu.com/ubuntu focal-updates/universe amd64 Packages [1,045 kB]
Get:39 http://us.archive.ubuntu.com/ubuntu focal-updates/universe Translation-en [246 kB]
Get:40 http://us.archive.ubuntu.com/ubuntu focal-updates/universe amd64 c-n-f Metadata [24.1 kB]
Get:41 http://us.archive.ubuntu.com/ubuntu focal-backports/main i386 Packages [36.1 kB]
Get:42 http://us.archive.ubuntu.com/ubuntu focal-backports/main amd64 Packages [45.7 kB]
Get:43 http://us.archive.ubuntu.com/ubuntu focal-backports/main Translation-en [16.3 kB]
Get:44 http://us.archive.ubuntu.com/ubuntu focal-backports/main amd64 c-n-f Metadata [1,420 B]
Hit:45 http://us.archive.ubuntu.com/ubuntu focal-backports/restricted amd64 c-n-f Metadata
Get:46 http://us.archive.ubuntu.com/ubuntu focal-backports/universe i386 Packages [13.8 kB]
Get:47 http://us.archive.ubuntu.com/ubuntu focal-backports/universe amd64 Packages [24.9 kB]
Get:48 http://us.archive.ubuntu.com/ubuntu focal-backports/universe Translation-en [16.3 kB]
Get:49 http://us.archive.ubuntu.com/ubuntu focal-backports/universe amd64 c-n-f Metadata [880 B]
Fetched 35.6 MB in 5s (7,087 kB/s)
Reading package lists... Done
2023-03-28T05:23:05-06:00 INFO Waiting for automatic snapd restart...
multipass 1.11.0 from Canonical✓ installed
OK
Get:1 https://download.docker.com/linux/ubuntu focal InRelease [57.7 kB]
Hit:2 http://us.archive.ubuntu.com/ubuntu focal InRelease
Hit:3 http://us.archive.ubuntu.com/ubuntu focal-updates InRelease
Get:4 https://download.docker.com/linux/ubuntu focal/stable amd64 Packages [24.9 kB]
Hit:5 http://us.archive.ubuntu.com/ubuntu focal-backports InRelease
Hit:6 http://security.ubuntu.com/ubuntu focal-security InRelease
Fetched 82.6 kB in 1s (121 kB/s)
Reading package lists... Done
Hit:1 https://download.docker.com/linux/ubuntu focal InRelease
Hit:2 http://us.archive.ubuntu.com/ubuntu focal InRelease
Hit:3 http://us.archive.ubuntu.com/ubuntu focal-updates InRelease
Hit:4 http://us.archive.ubuntu.com/ubuntu focal-backports InRelease
Hit:5 http://security.ubuntu.com/ubuntu focal-security InRelease
Reading package lists... Done
Cloning into '/tmp/kind'...
remote: Enumerating objects: 26177, done.
remote: Counting objects: 100% (474/474), done.
remote: Compressing objects: 100% (37/37), done.
remote: Total 26177 (delta 454), reused 437 (delta 437), pack-reused 25703
Receiving objects: 100% (26177/26177), 12.17 MiB | 20.33 MiB/s, done.
Resolving deltas: 100% (15008/15008), done.
go: downloading github.com/spf13/pflag v1.0.5
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/spf13/cobra v1.2.1
go: downloading github.com/alessio/shellescape v1.4.1
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/pelletier/go-toml v1.9.4
go: downloading gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
go: downloading github.com/BurntSushi/toml v0.4.1
go: downloading github.com/evanphx/json-patch/v5 v5.6.0
go: downloading sigs.k8s.io/yaml v1.3.0
go: downloading golang.org/x/sys v0.0.0-20210630005230-0f9fa26af87c
go: downloading gopkg.in/yaml.v2 v2.4.0
kernel.perf_event_paranoid = -1
net.netfilter.nf_conntrack_max = 655360
swapoff: cannot find the device for UUID=a96e550e-7190-4e25-995f-f1b4ed0d3648
```

</details>

<details> <summary> ./scripts/github_runner/start_runners.sh 1 https://github.com/curiousgeorgiy/vhive <GitHub PAT> integ </summary>

```console
ubuntu20base: Pulling from vhiveease/integ_test_runner
Digest: sha256:afd7c15323c48e3a4d36a89e457b309cec20852c586d1e5f29ef90b400725b03
Status: Image is up to date for vhiveease/integ_test_runner:ubuntu20base
docker.io/vhiveease/integ_test_runner:ubuntu20base
integration_test-github_runner-node-1.gh-705.ntu-cloud-pg0.utah.cloudlab.us-1
integration_test-github_runner-node-1.gh-705.ntu-cloud-pg0.utah.cloudlab.us-1
b621fbf7f9138b32ab0cfee7bda0d9ab515aad850d6402e0cbe7cf15023b5e9b
```

</details>

The scripts run successfully, but no runners are added to my repository.

My GitHub PAT scopes:
![Screenshot 2023-03-28 at 14 36 37](https://user-images.githubusercontent.com/32464072/228226038-0bcd4a1b-eb17-44de-9870-6c7f0724cd64.png)


[^1]: https://github.com/vhive-serverless/vHive/blob/main/docs/continuous_integration.md

ustiugov commented:
@CuriousGeorgiy @shyamjesal the question is how we build `vhiveease/integ_test_runner:ubuntu20base`

shyamjesal commented on docs/developers_guide.md:
````diff
@@ -106,6 +106,33 @@ We also offer self-hosted stock-Knative environments powered by KinD. To be able
+2. Setup the CRI test environment for the desired sandbox:
+
+```bash
+./scripts/github_runner/setup_cri_test_env.sh <sandbox>
````
Here as well.

CuriousGeorgiy replied:
Fixed.

shyamjesal commented on docs/developers_guide.md:
````diff
@@ -106,6 +106,33 @@ We also offer self-hosted stock-Knative environments powered by KinD. To be able
+4. Cleanup:
+
+```bash
+./scripts/github_runner/clean_cri_runner.sh <sandbox>
````
same here.

CuriousGeorgiy replied:
Fixed.

shyamjesal commented:
> @CuriousGeorgiy @shyamjesal the question is how we build `vhiveease/integ_test_runner:ubuntu20base`

Dockerfile.github_runner is the dockerfile for that.

ustiugov commented:
Then, we need to keep that docker image in the code base. @CuriousGeorgiy please revert.

Also, Georgiy, could you write up a Golang script `deploy_runners.go` that would take in IP addresses of VMs and start (or tear down and re-start upon a given option) runners, 2 VMs for `cri` gvisor and 2 VMs for `cri` firecracker (only 1 runner on each VM), and 1 VM for 6 `integ` runners concurrently in Docker (these run nicely in colocation). You can assume that provisioning of the VMs is done manually beforehand. 

> > @CuriousGeorgiy @shyamjesal the question is how we build `vhiveease/integ_test_runner:ubuntu20base`
> 
> Dockerfile.github_runner is the dockerfile for that.

CuriousGeorgiy commented:
@shyamjesal @ustiugov restored the Dockerfile for integration runners.

CuriousGeorgiy commented:
@ustiugov added a runner deployer script and added cron jobs for rebooting runner hosts to setup scripts.

ustiugov commented on docs/continuous_integration.md:
```diff
@@ -1,16 +1,34 @@
-./scripts/github_runner/start_runners.sh <num of runners> https://github.com/<OWNER>/<REPO> <Github Access key> integ
-# start cri test runners
-./scripts/github_runner/start_runners.sh <num of runners> https://github.com/<OWNER>/<REPO> <Github Access key> cri
+./scripts/github_runner/setup_bare_metal_runner.sh <GitHub organization> <Github PAT> <runner label> <runner name> <sandbox>
```
Please list the supported labels and sanboxes

CuriousGeorgiy replied:
Removed label parameter, added a list of sandboxes.

ustiugov replied:
where? did you push?

CuriousGeorgiy replied:
I pushed the wrong branch, should be up to date now.

ustiugov commented on docs/continuous_integration.md:
````diff
@@ -1,16 +1,34 @@
+3. Setup the desired number of runners:
+
+```bash
+./scripts/github_runner/setup_integ_runners.sh <num of runners> <GitHub Account> <Github PAT> [restart]
````
please specify `Github account` format and clarify what `restart` means

CuriousGeorgiy replied:
I don't know how to specify GitHub account... I believe it's self-explanatory. Like `CuriousGeorgiy` or `ustiugov` or `vhive-serverless`.

Added a comment about `restart` flag.

ustiugov replied:
I don't see the comment about restart, where?

just give these examples: vhive-serverless or your-username

CuriousGeorgiy replied:
I pushed the wrong branch, should be up to date now. Added example.

ustiugov commented on docs/continuous_integration.md:
```diff
@@ -1,16 +1,34 @@
-* Create Create GitHub personal access token for your repo with [these](https://github.com/myoung34/docker-github-actions-runner#create-github-personal-access-token) scopes.
+
+### Deploying integration tests workflow runners
+1. Create GitHub personal access token for your repo with [these](https://github.com/myoung34/docker-github-actions-runner#create-github-personal-access-token) scopes.
```
```suggestion
1. Create GitHub personal access token (PAT) for your repo with [these](https://github.com/myoung34/docker-github-actions-runner#create-github-personal-access-token) scopes.
```

CuriousGeorgiy replied:
Fixed.

ustiugov replied:
where? I don't see it. You use `GitHub PAT` below, hence you need to define it.

ustiugov reviewed (changes requested):
- [x] Rename the go script to `deploy_runners.go`
- [ ] The Go script's usage needs to be documented. Particularly the input JSON format. Also, please add a template JSON file for people to modify quickly. 
- [ ] copyrights of your authored files should have `your name and vHive team`.

CuriousGeorgiy commented:
@ustiugov 

> rename the go script to `deploy_runners.go`

Done.

> The Go script's usage needs to be documented. Particularly the input JSON format. Also, please add a template JSON file for people to modify quickly.

The runner deployer script usage is described in the CI guide, I also added a sample configuration file and added an example to the `conf` flag description.

> copyrights of your authored files should have `your name and vHive team`.

Done.

ustiugov commented on scripts/github_runner/setup_integ_runners.sh:
```diff
@@ -0,0 +1,85 @@
+API_HEADER="Accept: application/vnd.github.${API_VERSION}+json"
+AUTH_HEADER="Authorization: token ${GH_PAT}"
+
+_SHORT_URL="https://github.com/$2/vhive"
```
```suggestion
_SHORT_URL="https://github.com/$2/vHive"
```

CuriousGeorgiy replied:
Fixed.

ustiugov reviewed (approved):
Going to merge now but we'll test the CI fully right after
